pkg/reconciler/topic/topic.go

```diff
@@ -0,0 +1,418 @@
+				zap.Any("state", state),
+				zap.Error(err))
+
+			msg := "unknown"
+			if err != nil {
+				msg = err.Error()
+			}
+			topic.Status.MarkNoTopic(
+				"CreateFailed",
+				"Failed to create Topic: %q",
+				msg)
+			return err
+		}
+
+	case v1alpha1.TopicPolicyNoCreateNoDelete:
+		state, err := c.EnsureTopicExists(ctx, topic, topic.Spec.Project, topic.Status.TopicID)
+		// Check state.
+		switch state {
+		case pubsub.OpsJobGetFailed:
+			logger.Error("Failed to get topic ops job.",
+				zap.Any("propagationPolicy", topic.Spec.PropagationPolicy),
+				zap.Any("state", state),
+				zap.Error(err))
+			return err
+
+		case pubsub.OpsJobCreated:
+			// If we created a job to verify a topic, then update the status.
+			topic.Status.MarkTopicOperating("Verifying",
+				"Created Job to verify topic %q.",
+				topic.Status.TopicID)
+			return nil
+
+		case pubsub.OpsJobCompleteSuccessful:
+			topic.Status.MarkTopicReady()
+
+		case pubsub.OpsJobCreateFailed, pubsub.OpsJobCompleteFailed:
+			logger.Error("Failed to verify topic.",
+				zap.Any("propagationPolicy", topic.Spec.PropagationPolicy),
+				zap.Any("state", state),
+				zap.Error(err))
+
+			msg := "unknown"
+			if err != nil {
+				msg = err.Error()
+			}
+			topic.Status.MarkNoTopic(
+				"VerifyFailed",
+				"Failed to verify Topic: %q",
+				msg)
+			return err
+		}
+	default:
+		logger.Error("Unknown propagation policy.", zap.Any("propagationPolicy", topic.Spec.PropagationPolicy))
+		return nil
+	}
+
+	publisher, err := c.createOrUpdatePublisher(ctx, topic)
+	if err != nil {
+		logger.Error("Unable to create the publisher", zap.Error(err))
+		return err
+	}
+	topic.Status.PropagatePublisherStatus(publisher.Status.GetCondition(apis.ConditionReady))
+
+	if publisher.Status.IsReady() {
+		topic.Status.SetAddress(publisher.Status.Address.URL)
+	}
+	return nil
+}
+
+func (c *Reconciler) updateStatus(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, error) {
+	topic, err := c.topicLister.Topics(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(topic.Status, desired.Status) {
+		return topic, nil
+	}
+	becomesReady := desired.Status.IsReady() && !topic.Status.IsReady()
+	// Don't modify the informers copy.
+	existing := topic.DeepCopy()
+	existing.Status = desired.Status
+
+	ch, err := c.RunClientSet.PubsubV1alpha1().Topics(desired.Namespace).UpdateStatus(existing)
+	if err == nil && becomesReady {
+		duration := time.Since(ch.ObjectMeta.CreationTimestamp.Time)
+		c.Logger.Infof("Topic %q became ready after %v", topic.Name, duration)
+
+		if err := c.StatsReporter.ReportReady("Topic", topic.Namespace, topic.Name, duration); err != nil {
+			logging.FromContext(ctx).Infof("failed to record ready for Topic, %v", err)
+		}
+	}
+
+	return ch, err
+}
+
+func (c *Reconciler) updateFinalizers(ctx context.Context, desired *v1alpha1.Topic) (*v1alpha1.Topic, bool, error) {
```

Comment:
Add a comment saying "This method is generic for future compatibility with an SDK or some such"